### PR TITLE
[FrameworkBundle] allow configuring trusted proxies using semantic configuration

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | -
| Tickets       | -
| License       | MIT
| Doc PR        | -

On top of the improved DX this should provide, this PR (and #37351) will allow [removing the corresponding lines](https://github.com/symfony/recipes/pull/790) from `index.php` & recipes.

Using values:
```yaml
framework:
    trusted_proxies: '127.0.0.0/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16'
#or
    trusted_proxies: '%env(TRUSTED_PROXIES)%'
```

`trusted_headers` is similar but is an array of headers to trust.
```yaml
framework:
    # that's the defaults already
    trusted_headers: ['x-forwarded-all', '!x-forwarded-host', '!x-forwarded-prefix']
```

### CODE REVIEW

**Review by yceruto — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
PR is ready.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php [hunk 1]
```diff
@@ -120,6 +122,20 @@
...
+            ->args([
+                service('kernel'),
+                service('http_cache.store'),
+                service('esi')->nullOnInvalid(),
+                abstract_arg('options'),
+            ])
+            ->tag('container.hot_path')
+
+        ->set('http_cache.store', Store::class)
+        ->args([
```
Comment (maxhelias) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:135]: I think we should align args for readability 🙄

**Review by maxhelias — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
(rebased)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -242,6 +243,11 @@ public function load(array $configs, ContainerBuilder $container)
         $container->setParameter('kernel.default_locale', $config['default_locale']);
         $container->setParameter('kernel.error_controller', $config['error_controller']);
 
+        if (isset($config['trusted_proxies'], $config['trusted_headers'])) {
+            $container->setParameter('kernel.trusted_proxies', $config['trusted_proxies']);
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:247]: Can we explode here instead of at runtime? We might also want to trim?

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:247]: we can't really if we want to accept env vars here

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:247]: `trim()` added, at runtime: I tried, but we don't support csv-parsed env vars in array-node definitions. I'm sure that could be done, but that'd require a lot of work on the Config component before. For now at least, we need to support string+explode.

**Comment by nicolas-grekas** [comment]
Recipe update at https://github.com/symfony/recipes/pull/790

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by goetas** [comment]
Sorry for arriving this late to this pull request, but is this somehow related to the security issue in 3.3? (i'm talking about https://symfony.com/blog/fixing-the-trusted-proxies-configuration-for-symfony-3-3 )

**Comment by nicolas-grekas** [comment]
This PR builds on the knowledge gathered in the blog post and the related issues you linked. It doesn't reintroduce the original issue.
1. Trusted headers are configurable now, and have a safe default value (not trusting x-forwarded-host)
2. No proxies are trusted by default,
3. Runtime configuration happens very early when handling the request, earlier than HttpCache when it's enabled.
Did you spot an issue?

**Comment by goetas** [comment]
- from the DX point of view this change is much better, it removes a lot of "mysterious"  code from the `index.php` file
- from the implementation point of view it introduces even more coupling with the request object (in this case calling methods on it father than just blindly forwarding it to the handler).
 a better way could be having an event in the `preBoot` (but i guess the event dispatcher is not available yet), or a dedicated interface that can be used to hook into the boot process, and having  a dedicated class providing the feature of "configuring some global state". in this way other people get a way to hook by them self into the process..
but I guess that might be out of scope for this change

**Comment by nicolas-grekas** [comment]
You're correct. Providing extensibility here is a separate talk, that'd need a use case and someone to implement it. OSS dynamics: future will tell if/why and how we should did there.

**Comment by goetas** [comment]
I'm perfectly aware of how OSS works, it is just that symfony was often looking a bit more in the future and having static method calls here and there rarely has been a good idea.

**Comment by mpdude** [comment]
May I ask why the default is (and previously was, elsewhere) _not_ to trust `X-Forwarded-Host`? 

I understand that trusting this header has security implications, but if I am not misreading the change made here the prerequisite for evaluating the header is to configure a trusted proxy in the first place? And once you do that (because you have a proxy in place), wouldn't you want to also trust `X-Forwarded-Host`?

Also, on https://symfony.com/doc/master/deployment/proxies.html, the example configuration shown configures the proxy's IP address and uses `Request::HEADER_X_FORWARDED_ALL`. It does not even mentions _not_ trusting `X-Forwarded-Host`.

**Comment by nicolas-grekas** [comment]
@mpdude sure, check https://github.com/symfony/symfony/issues/20178

**Comment by mpdude** [comment]
Thank you Nicolas! 

Documentation has been re-arranged a bit since #20178, so I hope I did not miss anything that was more visible back then; also I followed the related doc PRs and hope I understood it all correctly.

We all 💯 that we must not trust any proxy or header by default.

David's intention back in #20178 was to make _trusting proxies_ more safe by default by _not_ having _any_ default headers that would be trusted once you configure proxy IPs. He was referring to AWS ELBs, that (to my knowledge, [to the present day](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/x-forwarded-headers.html) to not set `X-Forwarded-Host`. Not having any defaults should force people to think through their setup and choose the right set of headers.

Fabien [suggested](https://github.com/symfony/symfony/issues/20178#issuecomment-290908793) not to have defaults optimized for the AWS case.

I think the [current documentation](https://symfony.com/doc/current/deployment/proxies.html) is mostly in line with this: It describes to either provide the IP and use `Request::HEADER_X_FORWARDED_ALL`; or, for the AWS ELB case, to configure network security appropriately and then use `Request::HEADER_X_FORWARDED_AWS_ELB` in combination with trusting _all_ IPs.

The default configuration provided in the FrameworkBundle (with this PR) or [Flex/recipes in 5.1](https://github.com/nicolas-grekas/recipes/blob/b1e3b7f1e3c84ac2204c60dfa7c6209757e69940/symfony/framework-bundle/5.1/public/index.php) however is a bit odd, as it leans towards the AWS use case (`Request::HEADER_X_FORWARDED_ALL ^ Request::HEADER_X_FORWARDED_HOST` == `Request::HEADER_X_FORWARDED_AWS_ELB`).

So, for setups with an Apache-based Reverse Proxy (my current use case) you need to explicitly configure the IP + explicitly configure `x-forwarded-all`/`Request::HEADER_X_FORWARDED_ALL`. Probably the same for Varnish.

For AWS ELB, you need network-level security config outside the scope of Symfony + trust _all_ IPs.

Bottom line: 

It seems to me the defaults 
* are not right for the Apache/Varnish case
* the AWS ELB case requires careful extra steps anyway 
* are not a "more safe" than any other values (what about `X-Forwarded-For`, for example?)
* do not follow David's suggestion (to my understanding).

Maybe BC prevents us from changing this anyway, but what about going _without_  trusted headers or using `x-forwarded-all`?

**Comment by nicolas-grekas** [comment]
Side note: we often get security reports from ppl that missed that `trusted_hosts` exists.
But ppl are lazy (I' put myself in the bag). Not trusting any headers by default won't mean they'll think about all this. Instead, they'll copy/paste some blogpost/article. We all do this.
This means the solution is more in the documentation than in the code.
For the code part, this PR doesn't aim to change the preexisting default, precisely because it was not the topic.
If you want to lead an improvement on this, please open a separate issue.
